### DISCUSSION TITLE
What if a contract dies when it is holding assets?

### DISCUSSION THREAD

**Discussion by anilhelvaci** [Agoric/agoric-sdk#5469] [Q&A] [answered]
## What is the Problem Being Solved?
Say we are holding assets in a ZCFSeat that lives inside a contract. What happens to those funds if the contract dies?

## Description of the Design
It's unclear to me that what best solution would be. A way to persist those funds maybe?

**Comment by anilhelvaci** [comment]
I opened the issue as we discussed during the office hours @dckc

**Comment by dckc** [comment]
@Chris-Hibbert this question came up in discussion of #3226 pool-based loan protocol .
My understanding is: if a loan contract has assets in a seat as a result of handling some offers, and then it crashes, the assets are lost. Is that right?

If so, what are some ways to mitigate that risk? @dtribble @erights did you mention some _Heirs, Successors and Assigns_ technique in discussion recently?

@rowgraus @jeetraut fwiw: @anilhelvaci asked if solving this was a requirement for #3226 and I said no.

**Comment by Chris-Hibbert** [comment]
I haven't thought about solutions to this problem. *successors* sounds like an interesting approach, but I haven't heard more than the name. I hope to learn more.

**Comment by rowgraus** [comment]
@dckc is your understanding that this is not covered in @warner's current approach to contract upgrade?  Is the key difference here driven by the word 'dies' vs 'upgrades'?

@dtribble and I long ago talked about how to extend Zoe's offer safety to long lived contract-held assets like liquidity pools, but AFAIK it hasn't yet gotten out of the early design phase.

**Comment by dckc** [comment]
> @dckc is your understanding that this is not covered in @warner's current approach to contract upgrade?

Um... yes... I don't see how upgrade is relevant. If something in the upgrade approach helps, that would be a pleasant surprise.

> Is the key difference here driven by the word 'dies' vs 'upgrades'?

I suppose so.

**Comment by dckc** [comment]
This came up in a Zoe meeting yesterday and I think @erights answered it, but I don't clearly recall the answer. The meeting was recorded, so this date _should_ be enough info to find the recording.

**Comment by dckc** [comment]
@Tartuffo I have lost track of why this is MN-1. Do you remember why?

**Comment by Tartuffo** [comment]
I don't.  On the surface it sounds undesirable for the assets to be lost, but as someone referred to above, what does it mean for a contract to "die"?

**Comment by erights** [comment]
I assume, "die" -> "crash"

**Comment by Tartuffo** [comment]
@erights Does Zoe have any way to know if a contract has crashed, and thus be able to safely return the assets held in the seats, or do we need to "fix" this vulnerability for MN-1?

**Comment by michaelfig** [comment]
@Tartuffo please create an issue to track requirements and scheduling around this.

**Comment by Tartuffo** [comment, marked as answer]
The new issue for this is here: https://github.com/Agoric/agoric-sdk/issues/5486
